### ISSUE TITLE
Epic: reflock suggest - place the first pins automatically

### ISSUE THREAD

**Issue by a-grasso** [a-grasso/reflock#20]
## Summary

Pins work; nobody places them. `reflock check` at level 1 needs no adoption
effort, but level 2 requires a human to read every cross-reference and decide
what it claims about its target. The result is predictable: this repository has
**1 pin**, and the three other repositories surveyed have **0** between them -
1,665 pinnable references, one of them pinned.

This epic ships `reflock suggest`, an opt-in subcommand that does that pass
once, as a diff you read. It is an *onboarding* tool for an existing feature,
not a new checker.

A working prototype exists and has been validated on two repositories. This
issue records the accepted design and the order of work.

---

## How it decides

Two signals, each doing what the other cannot.

**Churn picks which references to pin.** A pin's entire cost is that it fires,
and it fires when its target unit changes. Rather than predicting that from the
prose, measure it: replay first-parent history and count commits where the
unit's fingerprint moved. First-parent matters - a change landing as both a
topic commit and a merge is one edit, and counting both overstates by ~2.3x.

Counts are smoothed, `(changes + 1) / (commits + 2)`. This was not premature
generalisation; the unsmoothed version failed on contact. Its first run picked
twelve vendored skill documents, all showing "0 changes", because they have
exactly one commit in their entire history. Churn 0 there means *never
observed*, not *stable*, so dead files beat quiet ones. With smoothing, an
unobserved file scores 0.33 and a file watched across fifty commits that moved
twice scores 0.06 - and the picks became real architecture decision records.

**A local [laya](https://github.com/NandhaKishorM/laya) model picks where each
pin binds.** Churn has no concept of a section. laya is a non-autoregressive
"System 1" decision encoder (Apache-2.0, ModernBERT-large) that answers typed
questions - `choice`, `score`, `noul` - in a single forward pass, with no text
generated and therefore nothing to parse or hallucinate. Anchor selection is a
textbook `choice`, and critically its options are rendered **per call**, so
every reference is scored against its own target's headings rather than a fixed
label set. Measured against the 192-token option head: median 6 headings (~41
tokens), max 16 (~172), zero targets over budget. It narrows

```diff
-See [ADR-0002](adr/0002-some-decision.md) for the rationale.
+See [ADR-0002](adr/0002-some-decision.md#decision) for the rationale.<!--@9ee09c8e-->
```

so the pin watches ~17 lines instead of ~90. That narrowing is what makes a
target quiet rather than merely dead, and it is why the model earns its place
at all (see "What was measured").

`none` is a first-class answer - roughly a quarter of references are about a
document as a whole and should stay unanchored. So is refusing the target's own
H1, which spans title to first heading and would pin the whole file while
reporting a section.

---

## What ships

**One CLI, not two.** The suggester has no standalone use: its entire output is
reflock syntax that only reflock's fingerprinting gives meaning to. It ships
here, as an opt-in extra.

The property that cannot be traded away: **`reflock check` stays stdlib-only**,
because it runs in every CI job on every commit. The checker is the product;
the suggester runs once, at adoption. The optional dependency is the whole
mechanism for having both.

```bash
pip  install reflock                       # stdlib only, unchanged
pip  install reflock[suggest]              # + the model runtime

brew install <tap>/reflock                 # unchanged
brew install <tap>/reflock-suggest         # + the model runtime
```

Homebrew has **no** optional dependencies - `depends_on => :optional` and
`--with-*` were removed in Homebrew 2.0 and are not returning, because bottles
cannot exist if the dependency set varies per install. So brew gets a companion
formula rather than an extra. Both resolve to the same subcommand;
`reflock-suggest` is a package name, never a command anyone types.

The model is **not** in the formula. Fetched on first `reflock suggest`,
verified by sha256, cached under `~/.cache`. A ~400 MB resource block in a tap
is miserable to update.

### Usage

```bash
reflock suggest .            # writes markers
reflock suggest . -n         # dry run
reflock stamp && reflock check
git diff                     # read it - this is a suggestion, not an authority
```

---

## Language: Python, settled

A rewrite in Go is off the table for as long as the suggester exists. laya is a PyTorch model with a ModernBERT tokenizer; running it from Go means
CGO wrappers around ONNX Runtime and HuggingFace tokenizers, or reimplementing
both. That side stays Python regardless.

Consequently the suggester imports `reflock_lib.engine` directly -
`build_index`, `parse_refs`, `resolve_target`, `unit_text`, `fingerprint`.
There is deliberately **no** serialization boundary between them. Same repo,
same release, same tests: a JSON contract between two modules of one package
would let the two definitions of "a unit" drift silently, where a direct import
breaks loudly at the signature. Keeping those definitions identical is the
entire correctness argument - a suggester that disagrees with the engine
produces pins the checker cannot see.

---

## Why it is a subcommand and not a plugin

The two meet in the markdown file and nowhere else. The suggester writes the
same characters a human would type, so:

- reflock needs no changes to accept its output;
- it cannot tell a suggested pin from a hand-typed one;
- the pins keep working if the suggester is uninstalled.

That also makes reflock the suggester's **test harness**. A hallucinated anchor
is a `DANGLING` finding, not a silent error, so the checker validates the
suggester as a side effect of its normal job. This is how anchor quality was
verified without labelling anything.

---

## What was measured

Two private repositories, anonymised here as **A** (~80 days of history, 635
pinnable references) and **B** (~14 days, 191 pinnable references). B was never
used for training and is the honest test.

**A frontier-model ceiling run** established the task is answerable at all:
43.5% of references judged pin-worthy, applied at high confidence, `stamp`
clean, `check` clean, and drift correctly detected on a deliberate edit. Anchor
suggestion resolved **55/55** - zero hallucinations, verified by reflock.

**A lexical baseline failed**: best F1 0.654 against 0.605 for
predict-everything-positive. Not a usable gate.

**laya zero-shot has no signal** - AUC 0.433 on pin-worthiness and
mean reciprocal rank 0.307 against 0.359 for random on anchor selection. Both
below their trivial baselines, so fine-tuning had to create the signal rather
than sharpen one.

**Fine-tuned laya on ~885 machine-generated labels** (no human labelling time),
held out by source file so near-duplicate references in one paragraph cannot
straddle the split:

| | zero-shot | fine-tuned | baseline |
|---|---|---|---|
| pin-worthiness AUC | 0.433 | 0.839 | 0.500 random |
| pin-worthiness F1 | 0.161 | 0.794 | 0.654 lexical |
| anchor top-1 | 0.080 | 0.653 | 0.293 always-`none` |
| anchor, on anchored cases only | 0.076 | 0.830 | - |

**But an ablation killed the pin-worthiness question.** Ranking the same
candidate set three ways, top 21 each:

| ranking | alarms/week |
|---|---|
| model only | 3.3 |
| churn only | **1.8** |
| model-positive + churn | **1.8** (identical selection) |

Churn does all the selection work. Shipping a 400 MB model to answer a question
`git log` answers as well is not defensible - so **the pin-worthiness question
was cut and churn replaced it**. The model survives for anchor selection only,
where churn cannot compete because it has no notion of a section. Note every
arm above still used the model's anchors; the ablation replaced selection, not
anchoring.

Two caveats recorded honestly:

- **Both test repositories are young** (~80 and ~14 days), so every alarm rate
  was measured during initial construction, when documentation churns hardest
  and pinning helps least. The measurements are not wrong, but they do not yet
  describe a mature repository.
- **Alarm rate is a cost metric, not a value metric.** It is minimised
  perfectly by pinning the deadest files in the repository. That is why churn
  cannot be the whole tool, and why the smoothing above matters more than it
  looks.

**Prototype end-to-end on B**, never trained on: 191 candidates → 15 pins
written → `reflock stamp` → `reflock check` reports **all references OK**, every
anchor resolved, zero parser disagreements.

---

## Work items

- [ ] **1. Close the pin-worthiness gap.** The tool currently pins the quietest
      references, not the ones making a claim, so it pins index tables and
      see-also rows. Harmless for alarm rate, wrong in meaning, and the first
      thing a reader of the diff notices. Needs a cheap filter, not the model
      that was cut - link text and sentence shape are probably enough.
- [ ] **2. Wire the ONNX int8 inference path.** A hard prerequisite for
      packaging, not an optimisation: ONNX Runtime is a real homebrew-core
      formula, torch is not packageable there, so the companion formula cannot
      exist until this lands. Export is already verified - **opset 18 is a
      floor** (at 17 the export completes and the runtime rejects the graph,
      because `Split` gains `num_outputs` only at 18), and laya's
      variable-option `[MASK]` gather agrees with torch to 1.5e-05 across 2, 13
      and 17 options. fp32
      weights are 1.6 GB of external data, so int8 is mandatory and the sha256
      check must cover both artifact files.
- [ ] **3. Move the suggester into this repository** as the `suggest`
      subcommand, imported lazily so nothing on `check`'s path touches it.
      Missing extra prints an install hint and exits non-zero.
- [ ] **4. Ship the `[suggest]` pip extra.**
- [ ] **5. Ship the `reflock-suggest` tap formula**, model fetched on first run.

### Deferred

- **Model distribution and licensing.** laya is Apache-2.0, so redistributing a
  fine-tuned checkpoint is permitted with attribution; the open questions are
  where the artifact is hosted and how it is versioned against reflock releases.
  The fine-tune is laya's own published GRPO recipe (proper scoring rule plus
  soft cross-entropy) with the multi-GPU parts removed.
- **A value metric.** Everything above steers by cost. The natural ground truth
  is co-change: when a target unit changed, was the referring file edited in the
  same commit? That yields true and false positives from history alone, needs no
  labels, and is the only way to answer "is this pin worth having" rather than
  "is this pin quiet". Worth its own issue.
- **Re-measurement on a mature repository**, per the caveat above.
